### ISSUE TITLE
FIPSEndpointStateEnabled option results in incorrect endpoint for us-gov-west-1 region

### ISSUE THREAD

**Issue by bjs-code** [aws/aws-sdk#268]
Confirm by changing [ ] to [x] below to ensure that it's a bug:
- [X] I've gone though [Developer Guide](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/welcome.html) and [API reference](https://docs.aws.amazon.com/sdk-for-go/api/)
- [X] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-go) for answers
- [X] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-go/issues) and didn't find any solution
  
**Describe the bug**
Incorrect FIPS endpoint chosen for us-gov-west-1 region.
This appears to have been introduced with https://github.com/aws/aws-sdk-go/pull/3938
We are attempting to use the `UseFIPSEndpoint` option introduced in the PR above, and targeting an ec2 endpoint in us-gov-west-1.  According to both the documentation, and `endpoints.json`, the correct endpoint should be.
`ec2.us-gov-west-1.amazonaws.com`
However, with the flag in place the endpoint chosen is:
`ec2-fips.us-gov-west-1.amazonaws.com` which doesn't resolve

**Version of AWS SDK for Go?**
v1.42.1

**Version of Go (`go version`)?**
go1.16.3 darwin/amd64

**To Reproduce (observed behavior)**
```
package main

import (
	"fmt"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/endpoints"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/ec2"
)

func main() {
	config := &aws.Config{
		Region:                        aws.String("us-west-gov-1"),
		CredentialsChainVerboseErrors: aws.Bool(true),
		UseFIPSEndpoint:               endpoints.FIPSEndpointStateEnabled,
	}

	sess := session.Must(session.NewSession(config))
	svc := ec2.New(sess)

	output, err := svc.DescribeImages(&ec2.DescribeImagesInput{
		ImageIds: []*string{aws.String("ami-12345678")},
	})

	fmt.Println(output, err)
}
```

**Expected behavior**
SDK uses correct endpoint, however
`go run main.go`
results in
```
{

} RequestError: send request failed
caused by: Post "https://ec2-fips.us-west-gov-1.amazonaws.com/": dial tcp: lookup ec2-fips.us-west-gov-1.amazonaws.com: no such host
```

**Additional context**

**Comment by seanorama** [comment]
I'm experiencing this same when using [packer](https://github.com/hashicorp/packer/) which uses [aws-sdk-go v1.42.29](https://github.com/hashicorp/packer/blob/main/go.mod#L137).

Failure:
```
$ export AWS_USE_FIPS_ENDPOINT=true

$ packer build
Error: Datasource.Execute failed: Error querying AMI: RequestError: send request failed
caused by: Post "https://ec2-fips.us-gov-west-1.amazonaws.com/": dial tcp: lookup ec2-fips.us-gov-west-1.amazonaws.com: no such host
```

Success:
```
$ export AWS_USE_FIPS_ENDPOINT=false

$ packer build
==> amazon-ebs.node: Prevalidating any provided VPC information
...
```

Evidence that the host doesn't exist:
```
$ host ec2-fips.us-gov-west-1.amazonaws.com
Host ec2-fips.us-gov-west-1.amazonaws.com not found: 3(NXDOMAIN)

$ host ec2.us-gov-west-1.amazonaws.com
ec2.us-gov-west-1.amazonaws.com has address 52.46.224.22
```

**Comment by seanorama** [comment]
Also seeing with terraform:

```
Error: RequestError: send request failed
caused by: Get "https://cloudfront-fips.us-west-2.amazonaws.com/...[SNIPPED]...": dial tcp: lookup cloudfront-fips.us-west-2.amazonaws.com on 172.17.64.15:53: no such host
```

**Comment by RanVaknin** [comment]
@seanorama @epatrice @jasdel @bjs-code @jasonmfehr 
Did you all get the endpoint wrong?

its `us-gov-west-1` **and not** `us-west-gov-1`
Consider this code:

```go
func main() {

	sess, err := session.NewSession(&aws.Config{
		Region:          aws.String("us-gov-west-1"),
		UseFIPSEndpoint: endpoints.FIPSEndpointStateEnabled,
	})
	if err != nil {
		log.Fatalf("failed to load config, %v", err)

	}
	svc := ec2.New(sess, aws.NewConfig().WithLogLevel(aws.LogDebugWithHTTPBody))
	
	_, out := svc.DescribeInstances(&ec2.DescribeInstancesInput{})

	fmt.Println(out)
}
```

output:

```console
---[ REQUEST POST-SIGN ]-----------------------------
POST / HTTP/1.1
Host: ec2.us-gov-west-1.amazonaws.com
User-Agent: aws-sdk-go/1.44.102 (go1.19.1; darwin; arm64)
Content-Length: 43
Authorization: REDACTED, SignedHeaders=content-length;content-type;host;x-amz-date, 
Signature=REDACTED
Content-Type: application/x-www-form-urlencoded; charset=utf-8
X-Amz-Date: 20221205T181624Z
Accept-Encoding: gzip

Action=DescribeInstances&Version=2016-11-15
```

I can see that the endpoint resolver works as expected. 
`us-gov-west-1` with FIPS enabled resolves to `ec2.us-gov-west-1.amazonaws.com` per the aws endpoint spec.
Same goes for S3 and SNS.

Am I misunderstanding the issue at hand?

Thank you,
Ran

**Comment by RanVaknin** [comment]
Hi @jasonmfehr @seanorama @epatrice @bjs-code 

I was able to confirm that EC2 has fixed their endpoints specification by running the same code with the old 1.42.1 version and was getting the **incorrect** endpoint, but with the current version of the SDK its working as expected.

S3, and SQS are working as expected as well.

Since this concern is addressed I feel confident we can close this issue. 
If any of you are experiencing this with other services, please open _**a separate issue**_ so we can upstream this with that service directly on a per-serivce-basis. 

Also, I highly recommend that you all migrate to Go SDK V2. We track the issues on that repo much more closely, and some of these endpoint concerns were addressed as part of a cross SDK endpoints related efforts.

Thank you,
Ran~

**Comment by seanorama** [comment]
@RanVaknin Thanks! As you saw in testing, we weren't setting the endpoint. We were only setting the variable and letting the SDK choose the endpoint.

Thanks for the fix!

**Comment by seanorama** [comment]
@RanVaknin When you say current version, do you mean the latest SDK v1 or to switch to SDK v2.

You may be aware that much of the community are reluctant to change to SDK v2 as it requires massive changes, in some cases complete re-architecture of applications.

**Comment by sbiscigl** [comment]
> When you say current version, do you mean the latest SDK v1 or to switch to SDK v2.

just chiming in because i happened to work on this with Ran yesterday. The latest version of V1 has the correct endpoints updated. specifically `v1.44.154` was tested. we tested the example code against `v1.44.154` and saw the correct endpoint returned, and verified that the wrong url was present in `v1.42.29`. I do believe any version > ` v1.44.113` will be fixed as i see the [endpoints.json file being updated that day for ec2](https://github.com/aws/aws-sdk-go/commit/e2962f353f41fe98b19be1b4b61e06fbe7c71a64). So updating V1 should resolve your issue.

**Comment by jonathan-hurley** [comment]
I had opened https://github.com/aws/aws-sdk-go/issues/4700 in the `aws-sdk-go` project since the latest version there still has this issue for several API endpoints (cloudformation, eks, etc). Essentially anywhere in the [AWS FIPS Compliance](https://aws.amazon.com/compliance/fips/) document where the GovCloud regions do **not** include the `-fips` modifier in the URL.

It feels like the endpoints should be corrected instead of the SDK. When FedRAMP auditors look at URLs in use and see those without the `-fips` modifier, they will flag them each and every time. 

But either way, still broken for `us-gov-east-1` and `us-gov-west-1` in the latest go SDK.

**Comment by RanVaknin** [comment]
Hi @jonathan-hurley,

Thanks for commenting here.

Like mentioned previously on this thread, the endpoint resolution for the SDKs is coming from a `endpoints.json` file. That file is generated based on each team's internal configuration and this is not something the SDK team can change.
We need to reach out to each affected team individually and have them fix their internal configuration for this to get reflected in the SDK -  since this is a Cross SDK & Cross Service team issue it will take a while to fix.

Some SDKs are moving into a new paradigm, decoupling the need for an upstream source. Go SDK is not one them (for now)

While we work on resolving this with each service team, a workaround would be to specify the correct endpoint manually:
```go
	sess, err := session.NewSession(&aws.Config{
		Region:          aws.String("us-gov-west-1"),
		UseFIPSEndpoint: endpoints.FIPSEndpointStateEnabled,
		Endpoint:        aws.String("https://cloudformation.us-gov-west-1.amazonaws.com"),
	})
```

>It feels like the endpoints should be corrected instead of the SDK. When FedRAMP auditors look at URLs in use and see those without the -fips modifier, they will flag them each and every time. 

In regards to the -fips flag being present in the resolved in the endpoint. I don't think this part will change, as you can still call that endpoint directly (not through the SDK) and changing this on the service level will likely **not** be backwards compatible.

I will contact the SA working with you and let them know when we have specific details for each service. 

In the meantime, it would be very helpful to know which other services you need a correct endpoint for so we can target those service teams first and unblock you. 

Thank you very much,
Ran~

**Comment by jonathan-hurley** [comment]
Hi @RanVaknin 

Unfortunately, the workaround specified would require a massive refactor as we use many different endpoints from a single session and is not feasible for our product. However, I understand your point about having each downstream team fix this since that's the current design of the SDK. 

With that said, it appears that this affects many endpoints. From what I can tell:
- this only affects GovCloud regions as non-Gov regions all have `-fips` in the URL
- some endpoints are OK for GovCloud since they already include `-fips` in the URL.

The services that we use which, from the documentation, seem to have a problem are:
- `monitoring`, `events`, `logs`, `autoscaler`, `eks`, `cloudformation`, `ec2`, `iam`, `kms`

I understand about not changing the documented URLs since it would cause backward compatibility problems. However adding new URLs with `-fips` in them would fix the issue and not cause and issues with existing users who use the URLs directly.

**Comment by seanorama** [comment]
> However adding new URLs with -fips in them would fix the issue and not cause and issues with existing users who use the URLs directly.

Definitely +1 to this.

1. It would allow the SDK to be simplified since the naming would be standard instead of human maintained lookups _(which are now >800K in Java SDK)._
2. It's an easy implementation (CNAME + add SAN to certificate) and common practice for situations like this.
3. Time lost due to confusion:
   1. Almost daily I'm asked _"why are you using some `-fips` and some not"_. To then send the person to an HTML page paying that some of those without `-fips` actually have FIPS..
   2. This is very unlike AWS where most things (like URLs) follow conventions and everything is available programmatically (APIs). Intsead, this is human maintained mappings in an HTML page.

**Comment by RanVaknin** [comment]
Since there is an active effort from all service team to address this, and there is no action item for the SDK team itself. I feel comfortable we can close this. 

If you have identified additional services that are impacted please let us know by opening a separate issue, and we can upstream that request internally to the service team.

Thank you!
Ran~
